File: solutions/typescript/0090.ts
/**
 * 90. 子集 II
 * 排序 + 回溯：同一层若 nums[i] === nums[i-1] 则跳过，
 * 保证对相同值只展开一次，避免重复子集。
 */
function subsetsWithDup(nums: number[]): number[][] {
    nums.sort((a, b) => a - b);
    const result: number[][] = [];
    const path: number[] = [];

    const backtrack = (start: number): void => {
        result.push([...path]);
        for (let i = start; i < nums.length; i++) {
            if (i > start && nums[i] === nums[i - 1]) continue;
            path.push(nums[i]);
            backtrack(i + 1);
            path.pop();
        }
    };

    backtrack(0);
    return result;
}

export default subsetsWithDup;

File: solutions/rust/0090.rs
/// 90. 子集 II
/// 排序 + 回溯：同一层若 nums[i] == nums[i-1] 则跳过，避免重复子集。
pub fn subsets_with_dup(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut nums = nums;
    nums.sort();
    let mut result: Vec<Vec<i32>> = Vec::new();
    let mut path: Vec<i32> = Vec::new();

    fn backtrack(
        nums: &[i32],
        start: usize,
        path: &mut Vec<i32>,
        result: &mut Vec<Vec<i32>>,
    ) {
        result.push(path.clone());
        for i in start..nums.len() {
            if i > start && nums[i] == nums[i - 1] {
                continue;
            }
            path.push(nums[i]);
            backtrack(nums, i + 1, path, result);
            path.pop();
        }
    }

    backtrack(&nums, 0, &mut path, &mut result);
    result
}

impl Solution {
    pub fn subsets_with_dup(nums: Vec<i32>) -> Vec<Vec<i32>> {
        subsets_with_dup(nums)
    }
}